drivers/clock_control/clock_control_nrfs_audiopll.c

```diff
@@ -0,0 +1,211 @@
+BUILD_ASSERT((int)AUDIOPLL_DIV_12 == (int)NRF_CLOCK_CONTROL_AUDIOPLL_DIV_12);
+BUILD_ASSERT((int)AUDIOPLL_DIV_16 == (int)NRF_CLOCK_CONTROL_AUDIOPLL_DIV_16);
+
+struct shim_data {
+	const struct device *dev;
+	struct k_sem evt_sem;
+	nrfs_audiopll_evt_type_t evt;
+};
+
+struct shim_config {
+	uint16_t freq_fraction;
+	enum nrf_clock_control_audiopll_prescaler_div div;
+};
+
+static void shim_nrfs_audiopll_evt_handler(nrfs_audiopll_evt_t const *evt, void *context)
+{
+	struct shim_data *dev_data = context;
+
+	dev_data->evt = evt->type;
+	k_sem_give(&dev_data->evt_sem);
+}
+
+static int shim_nrfs_request_enable(const struct device *dev)
+{
+	struct shim_data *dev_data = dev->data;
+	nrfs_err_t err;
+
+	LOG_DBG("send enable request");
+
+	err = nrfs_audiopll_enable_request(dev_data);
+	if (err != NRFS_SUCCESS) {
+		return -EAGAIN;
+	}
+
+	k_sem_take(&dev_data->evt_sem, K_FOREVER);
+	return dev_data->evt == NRFS_AUDIOPLL_EVT_ENABLED ? 0 : -EAGAIN;
+}
+
+static int shim_nrfs_request_disable(const struct device *dev)
+{
+	struct shim_data *dev_data = dev->data;
+	nrfs_err_t err;
+
+	LOG_DBG("send disable request");
+
+	err = nrfs_audiopll_disable_request(dev_data);
+	if (err != NRFS_SUCCESS) {
+		return -EAGAIN;
+	}
+
+	k_sem_take(&dev_data->evt_sem, K_FOREVER);
+	return dev_data->evt == NRFS_AUDIOPLL_EVT_DISABLED ? 0 : -EAGAIN;
+}
+
+static int shim_clock_control_api_on(const struct device *dev,
+				     clock_control_subsys_t sub_system)
+{
+	ARG_UNUSED(sub_system);
+
+	return shim_nrfs_request_enable(dev);
+}
+
+static int shim_clock_control_api_off(const struct device *dev,
+				      clock_control_subsys_t sub_system)
+{
+	ARG_UNUSED(sub_system);
+
+	return shim_nrfs_request_disable(dev);
+}
+
+static DEVICE_API(clock_control, shim_clock_control_api) = {
+	.on = shim_clock_control_api_on,
+	.off = shim_clock_control_api_off,
+};
+
+int nrf_clock_control_audiopll_set_freq(const struct device *dev, uint16_t freq_fraction)
+{
+	struct shim_data *dev_data = dev->data;
+	nrfs_err_t err;
+
+	LOG_DBG("send freq request");
+
+	err = nrfs_audiopll_request_freq(freq_fraction, dev_data);
+	if (err != NRFS_SUCCESS) {
+		return -EAGAIN;
+	}
+
+	k_sem_take(&dev_data->evt_sem, K_FOREVER);
+	return dev_data->evt == NRFS_AUDIOPLL_EVT_FREQ_CONFIRMED ? 0 : -EAGAIN;
+}
+
+int nrf_clock_control_audiopll_set_prescaler(const struct device *dev,
+					     enum nrf_clock_control_audiopll_prescaler_div div)
+{
+	struct shim_data *dev_data = dev->data;
+	nrfs_err_t err;
+
+	LOG_DBG("send prescaler request");
+
+	err = nrfs_audiopll_request_prescaler((enum audiopll_prescaler_div)div, dev_data);
+	if (err != NRFS_SUCCESS) {
+		return -EAGAIN;
+	}
+
+	k_sem_take(&dev_data->evt_sem, K_FOREVER);
+	return dev_data->evt == NRFS_AUDIOPLL_EVT_PRESCALER_CONFIRMED ? 0 : -EAGAIN;
+}
+
+int nrf_clock_control_audiopll_set_freq_inc(const struct device *dev,
+					    int8_t inc_val,
+					    uint16_t inc_period_ms)
```

Comment:
Seems that this API supports both increase and decrease. Suggest:

```suggestion
int nrf_clock_control_audiopll_set_freq_change(const struct device *dev,
					    int8_t change_val,
					    uint16_t change_period_ms)
```

Reply:
removed this API, periodic increase should just be done from the application if required, where the clock frequency is actually known. There is no callback or sync to NRFS so clock freq would immediately go out of sync if this method was used.